contracts/Insurance.sol

```diff
@@ -32,17 +34,219 @@ contract Insurance is IInsurance {
         tracer = ITracerPerpetualSwaps(_tracer);
         InsurancePoolToken _token = new InsurancePoolToken("Tracer Pool Token", "TPT");
         token = address(_token);
         collateralAsset = tracer.tracerQuoteToken();
+        // Start with 1, so you can set 0 as null equivalent
+        delayedWithdrawalCounter = 1;
+    }
+
+    /**
+     * @notice Commit to a delayed withdrawal, which will be executable in 5 days
+     * @param amount The amount of insurance pool tokens to burn
+     * @param extraScanIterations The user can volunteer to do extra expired withdrawal iterations
+     */
+    function commitToDelayedWithdrawal(uint256 amount, uint256 extraScanIterations) external override {
+        updatePoolAmount();
+        scanDelayedWithdrawals(SCAN_EXPIRED_WITHDRAWAL_COUNT + extraScanIterations);
+        uint256 balance = getPoolUserBalance(msg.sender);
+        require(balance >= amount, "INS: balance < amount");
+
+        clearUserDelayedWithdrawal(msg.sender);
+
+        InsurancePoolToken poolToken = InsurancePoolToken(token);
+
+        // tokens to return = (collateral holdings / pool token supply) * amount of pool tokens to withdraw
+        uint256 wadTokensToSend = LibInsurance.calcWithdrawAmount(
+            poolToken.totalSupply(),
+            publicCollateralAmount,
+            amount
+        );
+
+        uint256 fee = LibInsurance.calculateDelayedWithdrawalFee(
+            getPoolTarget(),
+            getPoolHoldings(),
+            totalPendingCollateralWithdrawals,
+            wadTokensToSend
+        );
+
+        // Tokens to send should be decreased by fee paid, so you do not have wadTokensToSend + fee > amount entitled
+        wadTokensToSend = wadTokensToSend - fee;
+        // fee can never be greater than publicCollateralAmount, since the `feeRate` in `feeRate * wadTokensToSend`
+        // will always be less than 1, meaning fee < wadTokensToSend.
+        // and wadTokensToSend <= publicCollateralAmount
+        publicCollateralAmount -= fee;
+        bufferCollateralAmount += fee;
+        totalPendingCollateralWithdrawals += wadTokensToSend;
+
+        uint256 id = addDelayedWithdrawal(amount, wadTokensToSend);
+        emit InsuranceDelayedWithdrawalCommit(address(tracer), msg.sender, wadTokensToSend, fee, id);
+    }
+
+    /**
+     * @notice Execute msg.sender's pending delayed withdrawal, if possible.
+     * @param extraScanIterations User can volunteer to scan through extra expired withdrawals
+     */
+    function executeDelayedWithdrawal(uint256 extraScanIterations) external override {
+        updatePoolAmount();
+        scanDelayedWithdrawals(SCAN_EXPIRED_WITHDRAWAL_COUNT + extraScanIterations);
+        uint256 balance = getPoolUserBalance(msg.sender);
+        uint256 id = accountsDelayedWithdrawal[msg.sender];
+        require(id != 0, "INS: No withdrawal pending");
+        require(!removeIfExpired(id), "INS: Withdrawal expired");
```

Comment:
Nit: This revert message is slightly misleading as the withdrawal may have been already executed. Optionally the check to see if a withdrawal commit has been executed already can be separated into a new function (outside of removeIfExpired) to both improve code clarity and revert message accuracy.